Token: Add operators and punctuators

Hywan opened the pull request:
Address https://github.com/tagua-vm/parser/issues/5, #8.
#### Specification

https://github.com/php/php-langspec/blob/master/spec/19-grammar.md#operators-and-punctuators
#### Progression
- [x] `[`,
- [x] `]`,
- [x] `(`,
- [x] `)`,
- [x] `{`,
- [x] `}`,
- [x] `.`,
- [x] `->`,
- [x] `++`,
- [x] `--`,
- [x] `**`,
- [x] `*`,
- [x] `+`,
- [x] `-`,
- [x] `~`,
- [x] `!`,
- [x] `$`,
- [x] `/`,
- [x] `%`,
- [x] `<<`,
- [x] `>>`,
- [x] `<`,
- [x] `>`,
- [x] `<=`,
- [x] `>=`,
- [x] `==`,
- [x] `===`,
- [x] `!=`,
- [x] `!==`,
- [x] `^`,
- [x] `|`,
- [x] `&`,
- [x] `&&`,
- [x] `||`,
- [x] `?`,
- [x] `:`,
- [x] `;`,
- [x] `=`,
- [x] `**=`,
- [x] `*=`,
- [x] `/=`,
- [x] `%=`,
- [x] `+=`,
- [x] `-=`,
- [x] `.=`,
- [x] `<<=`,
- [x] `>>=`,
- [x] `&=`,
- [x] `^=`,
- [x] `|=`,
- [x] `=&` (removed),
- [x] `,`,
- [x] `??`,
- [x] `<=>`,
- [x] `...`.

Hywan committed:
Token: Add operators and punctuators.

All the operators and punctuators from
https://github.com/php/php-langspec/blob/master/spec/19-grammar.md#operators-and-punctuators
are present. More operators, like `STATIC_CALL` or `TERNARY_ELSE`, have
been added for clarity reasons.

Hywan commented on source/tokens.rs:
```diff
@@ -58,6 +58,14 @@ token!(
 token!(
+    ADD: "+";
+    "The `ADD` token.\n\nRepresent the multiplication operator, e.g. `$x + $y`."
+);
```
Sure, the multiplication...

Hywan commented on source/tokens.rs:
```diff
@@ -114,6 +206,14 @@ token!(
+token!(
+    DIVIDE_AND_ASSIGN: "/=";
+    "The `DIVIDE_AND_ASSIGN` token.\n\nRepresent the division assignment operator, e.g. `$x /= $y;`."
+);
```
Remove last semicolon in the example.

Hywan commented on source/tokens.rs:
```diff
@@ -122,10 +222,22 @@ token!(
     "The `DOLLARS` token.\n\nRepresent the introduction of a variable, e.g. `$x`."
 );
 token!(
+    DOT: ".";
```
Remove this one.

Hywan commented on source/tokens.rs:
```diff
@@ -250,14 +414,34 @@ token!(
 );
 token!(
+    NOT: "!";
+    "The `NOT` token.\n\nRepresent the negation operator, e.g. `!$x`."
```
`BOOLEAN_NOT`

jubianchi commented on source/tokens.rs:
```diff
@@ -286,18 +470,50 @@ token!(
+    "The `TERNARY_ELSE` token.\n\nRepresent the falsy block of a ternary condition, e.g. `$x ? … : …`."
+);
+token!(
+    TERNARY_IF: "?";
```
`TERNARY_THEN` ?

Hywan replied:
I was so hesitant with this one… OK, so, `TERNARY_THEN`.

nikic replied:
Is it possible to define multiple tokens for one character? The chosen names often represent only one particular meaning, e.g. `?` could also represent a nullable type, `:` could be the delimiter for a return type or an alternative syntax, etc.

jubianchi replied:
@nikic yes we can. `token!` is just a simple macro we defined here. It allocate a string with a `static` lifetime. We can do things like:

``` rust
token!(
    TERNARY_THEN: "?";
    "The `TERNARY_THEN` token.\n\nRepresent the truly block of a ternary condition, e.g. `$x ? … : …`."
);

token!(
    NULLABLE_TYPE: "?";
    "The `NULLABLE_TYPE` token.\n\nblablabla"
);
```

But this would allocate the string twice. But this won't:

``` rust
token!(
    TERNARY_THEN: "?";
    "The `TERNARY_THEN` token.\n\nRepresent the truly block of a ternary condition, e.g. `$x ? … : …`."
);

token!(
    NULLABLE_TYPE: TERNARY_THEN;
    "The `NULLABLE_TYPE` token.\n\nblablabla"
);
```

But then, I find it less readable.

@Hywan we could change the `token!` macro to allow public/private tokens and do this:

``` rust
macro_rules! token {
    ($name:ident: $value:expr; $documentation:expr) => (
        #[doc=$documentation]
        const $name: &'static str = $value;
    );
    (pub $name:ident: $value:expr; $documentation:expr) => (
        #[doc=$documentation]
        pub const $name: &'static str = $value;
    );
}

token!(
    QUESTION_MARK: "?";
    "blablabla"
);

token!(
    TERNARY_THEN: QUESTION_MARK;
    "The `TERNARY_THEN` token.\n\nRepresent the truly block of a ternary condition, e.g. `$x ? … : …`."
);

token!(
    NULLABLE_TYPE: QUESTION_MARK;
    "The `NULLABLE_TYPE` token.\n\nblablabla"
);
```

Hywan replied:
Indeed!

jubianchi commented on source/tokens.rs:
```diff
@@ -286,18 +470,50 @@ token!(
+    "The `SUBSTRACT` token.\n\nRepresent the substraction operator, e.g. `$x - $y`."
+);
+token!(
+    SUBSTRACT_AND_ASSIGN: "-=";
```
`SUBTRACT`

Hywan replied:
w00t. “substract” is old-English, didn't know that. Thanks!

jubianchi replied:
@Hywan I didn't know either. I always used `subtract` :)

jubianchi commented on source/tokens.rs:
```diff
@@ -286,18 +470,50 @@ token!(
+    "The `STATIC_CALL` token.\n\nRepresent the static method call operator, e.g. `class::method()`."
+);
+token!(
+    SUBSTRACT: "-";
```
`SUBTRACT`

jubianchi commented on source/tokens.rs:
```diff
@@ -70,6 +78,70 @@ token!(
+    "The `ASSIGN` token.\n\nRepresent a binding of a value to a variable, e.g. `$x = 42`."
+);
+token!(
+    ASSIGN_REFERENCE: "=&";
```
Are you sure this should be a token ? It will be difficult to support notation like `= & $bla` with a single token.

Hywan replied:
I know. I was very hesitant with this one too. This is the grammar, but it does not mean there is no error. Also, this is not in the grammar of the Zend Engine. I will remove this one.

Hywan replied:
Note: This is probably for old PHP version.

nikic replied:
Removed in https://github.com/php/php-langspec/commit/854b2d1d01e5c3d3feec16eebf94f5481611a875.

Hywan replied:
Thanks @nikic!

jubianchi commented on source/tokens.rs:
```diff
@@ -238,10 +366,46 @@ token!(
+    "The `MULTIPLY` token.\n\nRepresent the multiplication operator, e.g. `$x * $y`."
+);
+token!(
+    MULTIPLY_AND_ASSIGN: "**=";
```
`*=`

Hywan replied:
Good catch!

Hywan committed:
Token: Remove `ASSIGN_REFERENCE`.

We can combine `ASSIGN` and `REFERENCE` to get this one, because
whitespaces can be present between the two.

See https://github.com/php/php-langspec/commit/854b2d1d01e5c3d3feec16eebf94f5481611a875.

Hywan committed:
Token: New `token!` variant with `pub`.

The goal is to be able to declare “private” tokens, in case their values
must be shared by public tokens.

Hywan committed:
Token: Add `FUNCTION_OUTPUT` and `NULLABLE`.

There are 2 new private tokens:

  * `COLON`, used by `FUNCTION_OUTPUT` and `TERNARY_ELSE`,
  * `QUESTION_MARK`, used by `NULLABLE` and `TERNARY_THEN`.

Hywan commented:
I have removed `PLUS` and `DOLLARS`, which are redundant with `ADD` and `VARIABLE`.
I have introduced the following private tokens: `COLON` and `QUESTION_MARK` (they are respectively used by `FUNCTION_OUTPUT` and `TERNARY_ELSE`, and by `NULLABLE` and `TERNARY_THEN`).
I have added `FUNCTION_OUTPUT` (didn't find a better name…) and `NULLABLE`.

Hywan commented:
I have rebased my work too. Ready for another review @jubianchi :smiley:!

Hywan committed:
Rule: Use the new `ADD` token.

Hywan commented:
ping @jubianchi?

jubianchi commented:
@Hywan good for me! Glab you chose to extends the `token!` macro, I think it will let us do some cool things!
